test/init_test.go

```diff
@@ -0,0 +1,183 @@
+// +build e2e
+
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains initialization logic for the tests, such as special magical global state that needs to be initialized.
+
+package test
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	knativetest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/tektoncd/pipeline/pkg/names"
+	"golang.org/x/xerrors"
+	yaml "gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+var initMetrics sync.Once
+
+func setup(t *testing.T) (*clients, string) {
+	t.Helper()
+	namespace := names.SimpleNameGenerator.RestrictLengthWithRandomSuffix("arrakis")
+
+	initializeLogsAndMetrics(t)
+
+	c := newClients(t, knativetest.Flags.Kubeconfig, knativetest.Flags.Cluster)
+	createNamespace(t, namespace, c.KubeClient)
+	verifyDefaultServiceAccountExists(t, namespace, c.KubeClient)
+	return c, namespace
+}
+
+func tearDown(t *testing.T, cs *clients, namespace string) {
+	t.Helper()
+	if cs.KubeClient == nil {
+		return
+	}
+	if t.Failed() {
+		header(t.Logf, fmt.Sprintf("Dumping objects from %s", namespace))
+		bs, err := getCRDYaml(cs, namespace)
+		if err != nil {
+			t.Error(err)
+		} else {
+			t.Log(string(bs))
+		}
+	}
+
+	t.Logf("Deleting namespace %s", namespace)
+	if err := cs.KubeClient.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{}); err != nil {
+		t.Errorf("Failed to delete namespace %s: %s", namespace, err)
+	}
+}
+
+func header(logf logging.FormatLogger, text string) {
+	left := "### "
+	right := " ###"
+	txt := left + text + right
+	bar := strings.Repeat("#", len(txt))
+	logf(bar)
+	logf(txt)
+	logf(bar)
+}
+
+func initializeLogsAndMetrics(t *testing.T) {
+	t.Helper()
+	initMetrics.Do(func() {
+		flag.Parse()
+		if err := flag.Set("alsologtostderr", "true"); err != nil {
+			t.Fatalf("Failed to set 'alsologtostderr' flag to 'true': %s", err)
+		}
+		logging.InitializeLogger(knativetest.Flags.LogVerbose)
+
+		if knativetest.Flags.EmitMetrics {
+			logging.InitializeMetricExporter(t.Name())
+		}
+	})
```

Comment:
if there is any way not to include this it would probably be good, it's not clear to me that we can even use the metrics exporting (i might be wrong tho!)

Reply:
The [plumbing](https://github.com/tektoncd/plumbing/blob/master/scripts/presubmit-tests.sh#L295) repo currently supports the `--emit-metrics` flag (see [here](https://github.com/tektoncd/plumbing/blob/master/scripts/presubmit-tests.sh#L295)). However, I don't know if anyone's using this option. @vdemeester do you know if the `--emit-metrics` option is being used?

Reply:
On `tektoncd` it is not being used I don't think

Reply:
Ok, thanks Vincent. I don't know if I should keep it then. Since the plumbing repo currently references an `--emit-metrics` flag, I'm leaning toward keeping it in. However, I don't feel strongly either way 😁

Reply:
> Metrics will be emitted for these Wait methods tracking how long test poll for.

(from pipeline's [`test/README`](https://github.com/tektoncd/pipeline/tree/master/test))
Hm, looks like the pipeline's `wait` library is [emitting metrics](https://github.com/tektoncd/pipeline/blob/master/test/wait.go#L76-L77). This data could be useful in the future, although I don't know if anyone's using this data right now.

Reply:
yes :+1: